### ISSUE TITLE
a little accuracy loss when running this code

### ISSUE THREAD

**Issue by cucdengjunli** [AlvinYH/Faster-VoxelPose#17]
![实验结果1](https://user-images.githubusercontent.com/46188762/208602927-9426c268-0942-4fb
![论文官方结果](https://user-images.githubusercontent.com/46188762/208604108-48df8cda-6285-43ce-ad1d-18890f7acef9.png)
6-a412-a9ca82b07696.png)

Dear authors:
It is grateful to read your paper and code. when i try to run this project to reproduce your paper work. **my result is dropped about 2mm**, could you explain why ?

is your code responde to this setting? using [5 views; mask; weights;].

my conda environment is that, show in the picture:

my GPU is RTX3090, cuda11.3 , torch1.11.0
![环境1](https://user-images.githubusercontent.com/46188762/208605017-cdf5a48e-a7f6-4f3d-bcd9-73094b7e1384.png)
![环境2](https://user-images.githubusercontent.com/46188762/208605024-b8f4385e-c93a-432c-acb2-aad6c6228b94.png)

**Comment by cucdengjunli** [comment]
this  is  my validation result on panoptic.  my training set is the same as    /Faster-VoxelPose-main/configs/panoptic/jln64.yaml

![实验结果1](https://user-images.githubusercontent.com/46188762/208605294-4b5c65d2-0e87-4dc8-b251-072921f0020c.png)

**Comment by cucdengjunli** [comment]
the backbone you provide is Resnet , the backbone you mentioned in paper is HRnet, maybe this is the reason. Let me exchange the backbone and see the result~

**Comment by cucdengjunli** [comment]
thanks for your perfect job！        **Could you please offer a HigherHRnet backbone Version？** 

the code maybe like this: 
https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation/blob/master/lib/models/pose_higher_hrnet.py
https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation/blob/master/experiments/coco/higher_hrnet/w32_512_adam_lr1e-3.yaml

**Comment by gpastal24** [comment]
Hi, how did you manage to train the model on RTX30 series gpu? Did you make any changes to the code?

**Comment by cucdengjunli** [comment]
> Hi, how did you manage to train the model on RTX30 series gpu? Did you make any changes to the code?

https://github.com/microsoft/voxelpose-pytorch/issues/19

I try this and succeed

**Comment by gpastal24** [comment]
> > Hi, how did you manage to train the model on RTX30 series gpu? Did you make any changes to the code?
> 
> [microsoft/voxelpose-pytorch#19](https://github.com/microsoft/voxelpose-pytorch/issues/19)
> 
> I try this and succeed

I did something similar, eventually. I returned the total loss at each iteration, and I got 18.6mm 3d error.

`       
        loss = loss_dict["total"]
        loss_2d = loss_dict["2d_heatmaps"]
        loss_1d = loss_dict["1d_heatmaps"]
        loss_bbox = loss_dict["bbox"]
        loss_joint = loss_dict["joint"]

        losses.update(loss.item())
        losses_2d.update(loss_2d.item())
        losses_1d.update(loss_1d.item())
        losses_bbox.update(loss_bbox.item())
        losses_joint.update(loss_joint.item())
        
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
`

**Comment by Mosh-Wang** [comment]
> > Hi, how did you manage to train the model on RTX30 series gpu? Did you make any changes to the code?
> 
> [microsoft/voxelpose-pytorch#19](https://github.com/microsoft/voxelpose-pytorch/issues/19)
> 
> I try this and succeed
Sorry, I would like to ask, did you use the method on the [https://github.com/microsoft/voxelpose-pytorch/issues/19](url) to change the loss part? Or what method did you use? Can you share it with me?

      loss_2d = loss_2d.mean()
      loss_3d = loss_3d.mean()
      loss_cord = loss_cord.mean()

      losses_2d.update(loss_2d.item())
      losses_3d.update(loss_3d.item())
      losses_cord.update(loss_cord.item())
      loss = loss_2d + loss_3d + loss_cord
      losses.update(loss.item())

      loss.backward()
      if (i + 1) % accumulation_steps == 0:
          optimizer.step()
          optimizer.zero_grad()

      # if loss_cord > 0:
      #     optimizer.zero_grad()
      #     (loss_2d + loss_cord).backward()
      #     optimizer.step()

      # if accu_loss_3d > 0 and (i + 1) % accumulation_steps == 0
      #     optimizer.step()
      #     optimizer.zero_grad()
      #     accu_loss_3d.backward()
      #     accu_loss_3d = 0.0
      # else:
      #     accu_loss_3d += loss_3d / accumulation_steps

      batch_time.update(time.time() - end)
      end = time.time()

**Comment by gpastal24** [comment]
@Mosh-Wang I just changed the code to backprop the total loss at every batch iteration. Loss dict is returned by the FVP model. I didn't do anything fancy.

https://github.com/AlvinYH/Faster-VoxelPose/blob/4daaedad466b9c95b1e9b35cfabd496b60e6013a/lib/models/voxelpose.py#L74-L80

**Comment by Mosh-Wang** [comment]
> @Mosh-Wang I just changed the code to backprop the total loss at every batch iteration. Loss dict is returned by the FVP model. I didn't do anything fancy.
> 
> https://github.com/AlvinYH/Faster-VoxelPose/blob/4daaedad466b9c95b1e9b35cfabd496b60e6013a/lib/models/voxelpose.py#L74-L80

Thank you very much for your reply. Can I ask you one more question? That is, when I am training Panoptic, I use TRAIN_HEATMAP_SRC: 'image' and TEST_HEATMAP_SRC: 'image' in the original config, and I get the following error. Do you also use this setting? Or have you changed it?
![ref](https://user-images.githubusercontent.com/56031726/229206495-f45326af-b03f-4173-82aa-c8b14a4cb952.PNG)

**Comment by Mosh-Wang** [comment]
> @Mosh-Wang
> 
> https://github.com/AlvinYH/Faster-VoxelPose/blob/4daaedad466b9c95b1e9b35cfabd496b60e6013a/lib/dataset/JointsDataset.py#L71
> 
> Change this to input_heatmaps or
> 
> https://github.com/AlvinYH/Faster-VoxelPose/blob/4daaedad466b9c95b1e9b35cfabd496b60e6013a/lib/dataset/JointsDataset.py#L168
> 
> to input_heatmap. They are not used anyway if you are using the image for training and testing. Just make a quick check before waiting for a whole epoch that this resolves the problem both for training and valdating

Thank you so much, it was so helpful! I can already successfully start training.
By the way, when you reproduce the result of Panoptic, did you set the part in the figure as true or false?
If pretrained backbone is used, does this part not need to be set to true?
![gra](https://user-images.githubusercontent.com/56031726/229215900-428c216b-f409-48fc-a0a4-0df26b6c0d4c.PNG)

**Comment by gpastal24** [comment]
@Mosh-Wang regarding the other q. I dont know if it matters that much. If you try both approaches would you br kind to let us know if training the 2D network as well, increases the performance of the method?

**Comment by zaie** [comment]
Maybe the omitted loss_off cause the performance drop.
https://github.com/AlvinYH/Faster-VoxelPose/issues/26

**Comment by cucdengjunli** [comment]
i reproduce the higherhrnet version backbone

**Comment by gpastal24** [comment]
@cucdengjunli 
> i reproduce the higherhrnet version backbone

Did you get the same results as the paper?

**Comment by AlvinYH** [comment]
Hi, @cucdengjunli. Thanks for your interest in our work. We've modified the code and you can pull the recent release. Yes, we make several changes to the model architecture (remove the offset branch and reduce the feature dimension in the weight_net). And the experimental results are slightly different from the one in the original paper. Specifically, as for Panoptic dataset, the mpjpe increases a little (+0.15mm) while the new model yields an improvement of 1.44 in terms of AP25. You can download the pre-trained checkpoints.

**Comment by cucdengjunli** [comment]
> @cucdengjunli
> 
> > i reproduce the higherhrnet version backbone
> 
> Did you get the same results as the paper?

yes , mpjpe@500mm: 17.966

**Comment by cucdengjunli** [comment]
> Hi, @cucdengjunli. Thanks for your interest in our work. We've modified the code and you can pull the recent release. Yes, we make several changes to the model architecture (remove the offset branch and reduce the feature dimension in the weight_net). And the experimental results are slightly different from the one in the original paper. Specifically, as for Panoptic dataset, the mpjpe increases a little (+0.15mm) while the new model yields an improvement of 1.44 in terms of AP25. You can download the pre-trained checkpoints. We'll revise our paper to specify these alternations. Also, thanks for pointing our mistake. We **did use Pose ResNet** for training on Panoptic Dataset instead of HigherHRNet. We'll fix this typo in the final version. And using HigherHRNet is expected to further reduce the errors.

thank you!

**Comment by CodeCrusader66** [comment]
> Hi, @cucdengjunli. Thanks for your interest in our work. We've modified the code and you can pull the recent release. Yes, we make several changes to the model architecture (remove the offset branch and reduce the feature dimension in the weight_net). And the experimental results are slightly different from the one in the original paper. Specifically, as for Panoptic dataset, the mpjpe increases a little (+0.15mm) while the new model yields an improvement of 1.44 in terms of AP25. You can download the pre-trained checkpoints. We'll revise our paper to specify these alternations. Also, thanks for pointing our mistake. We **did use Pose ResNet** for training on Panoptic Dataset instead of HigherHRNet. We'll fix this typo in the final version. And using HigherHRNet is expected to further reduce the errors.

I have alse reproduce the higherHRnet version code 😄,the result is as the same as your paper said. may I send you a merge request?

**Comment by 342600694** [comment]
> i reproduce the higherhrnet version backbone

Hello, may I ask how you modified HigherHRNet to train on the Panoptic dataset?

**Comment by JohnPekl** [comment]
I followed the README to run evaluation (inference, not training) and got a higher error MPJPE compared with yours; my output is `mpjpe@500mm: 57.998`. 

Download the [pretrained backbone model](https://drive.google.com/file/d/1Qlt_S5BoJcUIyXO5s6FEvVJ_ucQ1yN_L/view?usp=sharing)
Model weight: [Google drive](https://drive.google.com/file/d/1ETtd2isq11oiqqkEPfCQaOSmtU0MO9kH/view?usp=sharing)
Config: [cfg](https://github.com/AlvinYH/Faster-VoxelPose/blob/main/configs/panoptic/jln64.yaml)

Would you please let me know why?

```
=> the output will be saved in: output/panoptic/jln64
=> loading data...
Warning: detected an instance where the size of the bounding box is 2.24m, larger than 2m
=> 2580 images from 5 views loaded
=> constructing models...
=> save the 3D grid for feature sampling
=> loading weights of the backbone
=> load model state ../pretrained_weights/model_best.pth.tar

2025-07-17 18:37:18,067 => 2580 images from 5 views loaded
2025-07-17 18:37:20,645 => load model state ../pretrained_weights/model_best.pth.tar
2025-07-17 19:12:41,369 Evaluation results on Panoptic dataset:
ap@25: 0.0419	ap@50: 0.3421	ap@75: 0.7212	ap@100: 0.9260	ap@125: 0.9504	ap@150: 0.9573	recall@500mm: 0.9685	mpjpe@500mm: 57.998
```

**Comment by Sagalo04** [comment]
> > Hi, [@cucdengjunli](https://github.com/cucdengjunli). Thanks for your interest in our work. We've modified the code and you can pull the recent release. Yes, we make several changes to the model architecture (remove the offset branch and reduce the feature dimension in the weight_net). And the experimental results are slightly different from the one in the original paper. Specifically, as for Panoptic dataset, the mpjpe increases a little (+0.15mm) while the new model yields an improvement of 1.44 in terms of AP25. You can download the pre-trained checkpoints. We'll revise our paper to specify these alternations. Also, thanks for pointing our mistake. We **did use Pose ResNet** for training on Panoptic Dataset instead of HigherHRNet. We'll fix this typo in the final version. And using HigherHRNet is expected to further reduce the errors.
> 
> I have alse reproduce the higherHRnet version code 😄,the result is as the same as your paper said. may I send you a merge request?

@CodeCrusader66 
@cucdengjunli 

I would like to use the higherHRnet as a backbone pretrained on coco, however when I run the inference I got an error referring to the file lib/models/Higher_HRnet.py  which I don´t have. 

Could you please provide the modifications to reach it? Not only this file, also in other files. 
😄
